### PR TITLE
Audit fixes: Denom validation and multiHopSwap input check

### PR DESCRIPTION
**Pull request by NicholasDotSol**
validate token denoms
validate matching input tokens on multi-hop swap

neutron-integration-test-fixes: https://github.com/neutron-org/neutron-integration-tests/pull/299

### CODE REVIEW

File: x/dex/types/message_deposit.go [hunk 1]
```diff
@@ -69,6 +69,21 @@ func (msg *MsgDeposit) Validate() error {
...
+		return sdkerrors.Wrapf(ErrInvalidDenom, "TokenA denom (%s)", err)
+	}
+
+	err = sdk.ValidateDenom(msg.TokenB)
+	if err != nil {
+		return sdkerrors.Wrapf(ErrInvalidDenom, "TokenB denom (%s)", err)
+	}
+
+	if msg.TokenA == msg.TokenB {
+		return sdkerrors.Wrapf(ErrInvalidDenom, "tokenA == tokenB (%s)", err)
```
Comment (jcompagni10) [inline x/dex/types/message_deposit.go]: I don't think you mean to include `err` here

File: x/dex/types/message_multi_hop_swap.go [hunk 2]
```diff
@@ -70,17 +70,32 @@ func (msg *MsgMultiHopSwap) Validate() error {
 	if len(msg.Routes) == 0 {
 		return ErrMissingMultihopRoute
 	}
-
+	// set the expected tokenIn as the first token in the first route.
+	// if the route is empty, don't set tokenIn and let route length check error.
+	var tokenIn string
+	if len(msg.Routes[0].Hops) != 0 {
```
Comment (jcompagni10) [inline x/dex/types/message_multi_hop_swap.go]: We are already checking that len(msg.Routes[i].hopes) >= 2 below. I would refactor so we don't have to do this check twice. Also maybe better to keep this parallel with the `expectedExitToken` code below. May have to refactor both

File: x/dex/types/message_multi_hop_swap.go [hunk 3]
```diff
@@ -70,17 +70,32 @@ func (msg *MsgMultiHopSwap) Validate() error {
...
+		tokenIn = msg.Routes[0].Hops[0]
+	}
 	for _, route := range msg.Routes {
 		if len(route.Hops) < 2 {
 			return ErrRouteWithoutExitToken
 		}
 
+		if route.Hops[0] != tokenIn {
+			return sdkerrors.Wrapf(ErrInvalidDenom, "TokenIn mismatch in routes")
```
Comment (jcompagni10) [inline x/dex/types/message_multi_hop_swap.go]: Not really an `InvalidDenom` issue. I would either create a new error `ErrMultihopEntryTokensMismatch` or create a more general error that can be used for mismatch entry/exit tokens

File: x/dex/types/message_place_limit_order.go [hunk 4]
```diff
@@ -69,6 +69,21 @@ func (msg *MsgPlaceLimitOrder) Validate() error {
...
+		return sdkerrors.Wrapf(ErrInvalidDenom, "Error TokenIn denom (%s)", err)
+	}
+
+	err = sdk.ValidateDenom(msg.TokenOut)
+	if err != nil {
+		return sdkerrors.Wrapf(ErrInvalidDenom, "Error TokenOut denom (%s)", err)
+	}
+
+	if msg.TokenIn == msg.TokenOut {
+		return sdkerrors.Wrapf(ErrInvalidDenom, "tokenIn == tokenOut (%s)", err)
```
Comment (jcompagni10) [inline x/dex/types/message_place_limit_order.go]: remove `err`?

File: x/dex/types/message_withdrawal.go [hunk 5]
```diff
@@ -62,6 +62,21 @@ func (msg *MsgWithdrawal) Validate() error {
...
+		return sdkerrors.Wrapf(ErrInvalidDenom, "TokenA denom (%s)", err)
+	}
+
+	err = sdk.ValidateDenom(msg.TokenB)
+	if err != nil {
+		return sdkerrors.Wrapf(ErrInvalidDenom, "TokenB denom (%s)", err)
+	}
+
+	if msg.TokenA == msg.TokenB {
+		return sdkerrors.Wrapf(ErrInvalidDenom, "tokenA == tokenB (%s)", err)
```
Comment (jcompagni10) [inline x/dex/types/message_withdrawal.go]: remove `err` ?

**Review by jcompagni10 — CHANGES_REQUESTED** [review]

File: x/dex/types/message_multi_hop_swap.go (hunk 2 above)
Comment (NicholasDotSol) [reply x/dex/types/message_multi_hop_swap.go]: agree this function is getting ugly. will refactor

File: x/dex/keeper/msg_server_test.go [hunk 6]
```diff
@@ -2158,20 +2311,48 @@ func TestMsgMultiHopSwapValidate(t *testing.T) {
...
+					{Hops: []string{"TokenA", "TokenB", "TokenC"}},
+					{Hops: []string{"TokenA", "TokenB", "TokenD", "TokenE", "er", "TokenC"}},
+				},
+				AmountIn:       sdkmath.OneInt(),
+				ExitLimitPrice: math_utils.MustNewPrecDecFromStr("0.9"),
+			},
+			types.ErrInvalidDenom,
+		},
+		{
+			"taker denom mismatch in route",
```
Comment (jcompagni10) [inline x/dex/keeper/msg_server_test.go]: I would call this "maker denom" or "entry token"

**Review by jcompagni10 — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9308510496

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9310625760

**Review by pr0n00gler — CHANGES_REQUESTED** [review]
Integration tests did not pass, please fix them

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9319434742

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**